[cueweb] CueWeb system: First web-based release of CueGUI with many features from Cuetopia

ramonfigueiredo opened the pull request:
CueWeb is a web-based application that brings the core functionality of [CueGUI](https://github.com/AcademySoftwareFoundation/OpenCue/tree/master/cuegui), including Cuetopia and Cuecommander, to a web-accessible format. This initial version includes most Cuetopia features, with Cuecommander integration planned for the next phase. CueWeb simplifies rendering job management with customizable job tables, advanced filtering, detailed inspections, log viewing, and light/dark mode toggling, making it efficient and accessible across platforms. Finally, CueWeb leverages the [OpenCue REST Gateway](https://github.com/AcademySoftwareFoundation/OpenCue/tree/master/rest_gateway) to provide a REST endpoint for seamless interaction with the OpenCue gRPC API. 

CueWeb was developed using [React.js](https://react.dev/), [Next.js](https://nextjs.org/), [ShadCN UI](https://ui.shadcn.com/), and [NextAuth.js](https://next-auth.js.org/) for authentications. Authentication Providers in NextAuth.js are services that can be used to sign in to a user.

There are four ways a user can be signed in:

- [Using a built-in OAuth Provider](https://next-auth.js.org/configuration/providers/oauth) (e.g Github, Google, Okta, Apple, GitLab, Amazon, Microsoft Azure, LinkedIn, Atlassian, Auth0, etc...)
   - List of NextAuth.js providers: https://next-auth.js.org/providers/.
- [Using a custom OAuth Provider](https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider)
- [Using Email](https://next-auth.js.org/configuration/providers/email)
- [Using Credentials](https://next-auth.js.org/configuration/providers/credentials)

Core features and code changes:
- User authentication:
   - Secure login capabilities. Initial version using Okta, Google, and GitHub.
   - Added functionality to make login usage optional in CueWeb. See cueweb/README.md for more information.

- Job management dashboard:
  - Customizable table views to show or hide specific columns.
  - Filter jobs by state (active, paused, or completed).
  - Monitor or unmonitor jobs across various statuses.
  - Detailed job inspection via pop-ups displaying associated layers and frames.
  - Frame navigation with hyperlinks to logs and data pages.

- Job search functionality:
   - Search for jobs using show names followed by a hyphen.
   - Dropdown suggestions for matching jobs based on naming conventions like show1-shot-test_job_123.
   - Regex-enabled search (triggered by !) for advanced query patterns, with a tooltip for guidance.

- Dark mode toggle: 
   - Allows users to switch between light and dark themes.

- Optimized search and results loading:
   - Virtualized lists via FixedSizeList to improve performance.
   - Web worker implementation for filtering, reducing main thread workload.
   - Loading animations and efficient API call handling.

- Multi-job management:
   - Add or remove multiple jobs from the dashboard directly from search results.
   - Highlight jobs already in the table with a green indicator.

- Actions and context menu
   - Added actions for jobs, layers, and frames, including pause, unpause, retry, kill, and eat dead frames.
   - Option to un-monitor all/selected/finished/paused jobs in the jobs data table.
   - Context-sensitive menus disable options for completed jobs and ensure proper screen rendering.

- Table auto-reloading: 
   - All tables (jobs, layers, frames) now update at regular intervals for real-time data.

- View frame logs:
   - View previous and current log versions with a dropdown menu for selection.

- Authorization and security:
   - Authorization headers included in all REST gateway requests.
   - JWT tokens generated securely for API authentication.

- Unit testing:
   - Added Jest tests for authentication middleware, error handling, and JWT creation.

- Include some CueWeb screenshots examples: https://github.com/AcademySoftwareFoundation/OpenCue/blob/master/cueweb/interface_screenshots

**Link the Issue(s) this Pull Request is related to.**
Create the first version of the CueWeb system: a web-based application that replicates the core functionality of CueGUI: https://github.com/AcademySoftwareFoundation/OpenCue/issues/1595


Co-authored-by: Mariz Fahmy <mfahmy@imageworks.com>
Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Tomi Lui <tlui@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

marizf888 committed:
CueWeb system (#1356)

* CueWeb system (first version): The CueWeb system offers a robust set of features designed to enhance user interaction and productivity:

- User authentication: Secure login capabilities are supported through Okta, Google (Gmail), and GitHub.
- Job management dashboard: Upon logging in, users are greeted with a data table displaying their jobs. Features of this dashboard include:
  - Customizable views: Users have the ability to choose which columns are visible in the job table.
  - Job filtering: The table can be filtered based on the job state, allowing users to easily sort through active, paused, or completed jobs.
  - Job monitoring control: Users can unmonitor jobs whether they are selected, paused, finished, or include all job statuses.
  - Detailed job inspection: Clicking a button at the end of each row opens a detailed pop-up window showing layers and frames associated with that job.
  - Frame navigation: Each row in the frames table includes a hyperlink that directs to a dedicated page for that frame's data and logs, if available.
- Job search functionality: Users can search for specific jobs to monitor by typing a show name followed by a hyphen. This action triggers a dropdown menu displaying matching job entries, which can then be added to the user's table.
  - Naming convention: Jobs are expected to follow a naming pattern such as `showname-` followed by additional identifiers, for example, `show1-shot-fake_test_job_name_123456789`.
- Dark mode toggle: Users can switch between light and dark modes according to their preference.

Co-authored-by: Mariz Fahmy <mfahmy@imageworks.com>
Co-authored-by: Tomi Lui <tlui@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

* CueWeb system (first version): The CueWeb system offers a robust set of features designed to enhance user interaction and productivity:

- User authentication: Secure login capabilities are supported through Okta, Google (Gmail), and GitHub.
- Job management dashboard: Upon logging in, users are greeted with a data table displaying their jobs. Features of this dashboard include:
  - Customizable views: Users have the ability to choose which columns are visible in the job table.
  - Job filtering: The table can be filtered based on the job state, allowing users to easily sort through active, paused, or completed jobs.
  - Job monitoring control: Users can unmonitor jobs whether they are selected, paused, finished, or include all job statuses.
  - Detailed job inspection: Clicking a button at the end of each row opens a detailed pop-up window showing layers and frames associated with that job.
  - Frame navigation: Each row in the frames table includes a hyperlink that directs to a dedicated page for that frame's data and logs, if available.
- Job search functionality: Users can search for specific jobs to monitor by typing a show name followed by a hyphen. This action triggers a dropdown menu displaying matching job entries, which can then be added to the user's table.
  - Naming convention: Jobs are expected to follow a naming pattern such as `showname-` followed by additional identifiers, for example, `show1-shot-fake_test_job_name_123456789`.
- Dark mode toggle: Users can switch between light and dark modes according to their preference.

Co-authored-by: Mariz Fahmy <mfahmy@imageworks.com>
Co-authored-by: Tomi Lui <tlui@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

---------

Co-authored-by: Tomi Lui <tlui@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

Zach-Fong committed:
[cueweb] Enhance CueWeb search functionality (#1439)

[cueweb] Enhance CueWeb search functionality

Features and improvements:

1) Regex search implementation

- Added regex search capability: Use '!' after queries to enable regex search, which passes a different argument to the same REST endpoint as normal searches.
- Included a tooltip explaining how to use regex queries when hovering over the search box.

2) Improved results loading

- Optimized loading times by replacing `cmdk Command` components like list with `FixedSizeList` to enable virtualization.
- Limited API calls to trigger only when the show name in the query changes.
- Introduced a web worker to handle result filtering on an additional thread, previously managed by `cmdk list`.
- Added a loading animation for better user experience during API queries and result filtering.

3) Multi-job add/remove feature

- Enabled adding/removing multiple jobs to/from the jobs table directly from the search results dropdown.
- Highlighted jobs already present in the table with green for easier identification.

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

Zach-Fong committed:
[cueweb] CueWeb improvements and add unit testing (#1457)

Features and improvements

1. CueWeb authorization updates
- Modified CueWeb to include authorization headers in all HTTP requests
to the gRPC REST gateway.
- Added functionality to generate JWT tokens using a secret for
inclusion in authorization headers.
2. Include unit testing
- Introduced Jest tests for CueWeb to cover: Authentication middleware,
error handling, and JWT creation.

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

Zach-Fong committed:
[cueweb] Actions, context menu, auto-reload, and search by show-shot (#1578)

Features and improvements:

1. Actions

* Added the ability to trigger the following actions for selected jobs in the data table:
  * Eat dead frames
  * Retry dead frames
  * Pause
  * Unpause
  * Kill

2. Context menu

* Added a context menu that allows users to trigger actions for jobs, layers, and frames
* Job actions:
  * Unmonitor
  * Pause
  * Retry dead frames
  * Eat dead frames
  * Kill
* Layer actions:
  * Kill
  * Eat
  * Retry
  * Retry dead frames
* Frame actions:
  * Retry
  * Eat
  * Kill
* Menu items are disabled and greyed out if the job selected has finished
* The context menu will always open onscreen and will not render off the page

3. Table auto-reloading

* All tables (jobs, layers, and frames) will now query for their data on regular intervals to show the latest attributes of the object

4. Search by show-shot

* Searching is now changed to require search by `show-shot-` instead of search by `show-`, reducing the number of results returned

5. View previous logs

* Previous log versions are now viewable on the frames page with a dropdown that allows users to select older logs

6. Fix table selection
* Fixed the jobs table to index selected jobs using a jobs id instead of the index position
* Fixed issues where removed jobs were improperly unselected

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

Zach-Fong committed:
[cueweb] Prevent page reload on data updates in jobs, layers, and frames tables (#1580)

- Fixed an issue where data updates caused the page index to reset in the jobs, layers, and frames tables.
- Added type check in Jobs worker to verify received data format.

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

Zach-Fong committed:
[cueweb] Add autoload user jobs, fix frames page, and update interface screenshots (#1592)

- Added functionality that enables users to autoload their jobs
- Fixed the frames page to display the frames table properly
- Updated interface screenshots
- Added fixes for components that didn't work in dark mode

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

ramonfigueiredo committed:
[CueWeb] Update README.md

ramonfigueiredo committed:
[CueWeb] Update CueWeb Dockerfile

ramonfigueiredo committed:
[CueWeb] Code refactoring
- Remove extra spaces in the import statement of action_utils.test.ts

ramonfigueiredo commented:
Below are some screenshots of the CueWeb

## Log in page

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/0623cae7-f69f-4df0-ba5e-ee3cd55e8e2d">

## Main page (light mode = default)

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/44d4d134-2271-4c1f-bf9a-e4694c87b03d">

## Main page (dark mode)

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/95a67602-0aec-4e95-b9c3-6f7f1bcda12f">

## Layers / Frames page (light mode) 

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/bf28ef3a-b683-465e-afdf-f00892ae2cf9">

## Layers / Frames page (dark  mode) 

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/3794d842-afe2-4a97-a1d7-f20d5ce2f1a9">

## Frames logs page (light mode) 

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/0d24b3f1-1a9e-429c-ab6b-c04137d9f0f2">


## Frames logs page (dark mode) 

<img width="1603" alt="image" src="https://github.com/user-attachments/assets/a212ef61-f04a-4a95-a8bf-98c9b27b4250">

ramonfigueiredo commented:
Hi @DiegoTavares and @bcipriano 

The CueWeb code is ready for your review and to be merged into the master.

I already did many reviews throughout the year (from April 2024 to November 2024) and it is working fine inside the SPI infrastructure.

Thanks!

ramonfigueiredo committed:
[CueWeb] Add new images to the cueweb/interface_screenshots

ramonfigueiredo committed:
[CueWeb] Update the cueweb/README.md file with NextAuth.js's authentication providers' information

ramonfigueiredo commented:
FYI ... 
@tomi-lui @marizf888 @Zach-Fong

marizf888 commented:
> FYI ... @tomi-lui @marizf888 @Zach-Fong

Super happy to see CueWeb getting improved and more features being added 😁

ramonfigueiredo committed:
[CueWeb] Add new images to the cueweb/interface_screenshots

DiegoTavares commented on cueweb/.env.example:
```diff
@@ -0,0 +1,23 @@
+
+SENTRY_ENVIRONMENT='development'
+
+
```
It's worth adding a comment saying not all options are required and that if none of them are configured, the feature is disabled.

ramonfigueiredo replied:
Done!

See commit: [cueweb] Make CueWeb authentication and Sentry usage optional
https://github.com/AcademySoftwareFoundation/OpenCue/pull/1596/commits/9c326b6b4e072563281fa3a92ac8cbd8c93f6ef3

- Added functionality to make login and Sentry usage optional in CueWeb.
- Updated configuration to enable/disable these features via environment variables.
- Enhanced the cueweb/README.md with instructions for configuring these options in the CueWeb setup.

### About Sentry usage in CueWeb:

- Sentry environment variables
   - If you use the [Sentry](https://sentry.io/) system to monitor the CueWeb system, the following environment variables should be set:
      - SENTRY_ENVIRONMENT
      - SENTRY_URL
      - SENTRY_DSN
      - SENTRY_ORG
      - SENTRY_PROJECT
   - If you do not want to use Sentry, do not define `SENTRY_DSN`.

Change `.env` 

From:
```env
# Sentry values
SENTRY_ENVIRONMENT='development'
SENTRY_DSN = sentrydsn
SENTRY_URL = sentryurl
SENTRY_ORG = sentryorg
SENTRY_PROJECT = sentryproject
```

To:

```env
# Sentry values
SENTRY_ENVIRONMENT='development'
# SENTRY_DSN = sentrydsn
SENTRY_URL = sentryurl
SENTRY_ORG = sentryorg
SENTRY_PROJECT = sentryproject
```

DiegoTavares commented on cueweb/README.md:
```diff
@@ -0,0 +1,324 @@
+    - NEXT_JWT_SECRET
+        - This is used to create a JWT token which is required to access the REST endpoint of the opencue gRPC API
+
+    - Authentication environment variables
+        - Depending on which provider you use for authentication, you may have to set certain environment variables. 
+        - For example, for authentication (see file `cueweb/lib/auth.ts`), the following environment variables must be set:
+            - `NEXT_PUBLIC_AUTH_PROVIDER=github,okta,google` will show the three authentication buttons (Okta, Google, GitHub).
+            - `NEXT_PUBLIC_AUTH_PROVIDER=github,google` will show the two authentication buttons (GitHub and Google).
+            - `NEXT_PUBLIC_AUTH_PROVIDER=google` will show only the Google (Gmail) authentication button.
+
+          - [Okta](https://www.okta.com/)
+              - NEXT_AUTH_OKTA_CLIENT_ID
+              - NEXT_AUTH_OKTA_ISSUER
+              - NEXT_AUTH_OKTA_CLIENT_SECRET
+          - Google (Gmail)
+            - GOOGLE_CLIENT_ID
+            - GOOGLE_CLIENT_SECRET
+          - GitHub
+            - GITHUB_ID
+            - GITHUB_SECRET
```
As mentioned on yesterday's meeting. For some studios, having authentication can be a blocker. I think this section should explain how to turn authentication off.

ramonfigueiredo replied:
Done!

See commit: [cueweb] Make CueWeb authentication and Sentry usage optional
https://github.com/AcademySoftwareFoundation/OpenCue/pull/1596/commits/9c326b6b4e072563281fa3a92ac8cbd8c93f6ef3

- Added functionality to make login and Sentry usage optional in CueWeb.
- Updated configuration to enable/disable these features via environment variables.
- Enhanced the cueweb/README.md with instructions for configuring these options in the CueWeb setup.

### About CueWeb authentication usage

- Authentication environment variables
   - Depending on which [Next-Auth.js](https://next-auth.js.org/) provider you use for authentication, you may have to set certain environment variables. 
   - For example, for authentication (see file `cueweb/lib/auth.ts`), the following environment variables must be set:
      - `NEXT_PUBLIC_AUTH_PROVIDER=github,okta,google` will show the three authentication buttons (Okta, Google, GitHub).
      - `NEXT_PUBLIC_AUTH_PROVIDER=github,google` will show the two authentication buttons (GitHub and Google).
      - `NEXT_PUBLIC_AUTH_PROVIDER=google` will show only the Google (Gmail) authentication button.

      - [Okta](https://www.okta.com/)
            - NEXT_AUTH_OKTA_CLIENT_ID
            - NEXT_AUTH_OKTA_ISSUER
            - NEXT_AUTH_OKTA_CLIENT_SECRET
      - Google (Gmail)
            - GOOGLE_CLIENT_ID
            - GOOGLE_CLIENT_SECRET
      - GitHub
            - GITHUB_ID
            - GITHUB_SECRET

- To disable the CueWeb authentication, do not define `NEXT_PUBLIC_AUTH_PROVIDER`.

Change `.env`

From:

```env
# Authentication Configuration:
NEXT_PUBLIC_AUTH_PROVIDER=github,okta,google
```

To:

```env
# Authentication Configuration:
# NEXT_PUBLIC_AUTH_PROVIDER=github,okta,google
```

#### Authentication Setup 

The CueWeb project utilizes the [NextAuth.js](https://next-auth.js.org/) library for authentication, which includes many popular providers out-of-the-box for additional configuration or for implementing your own email authentication with a custom database. This project already implements Google, Github and Okta authentication and they are enabled if their respective environment variables are provided, otherwise ignored. 

Authentication providers in [NextAuth.js](https://next-auth.js.org/) are services that can be used to sign in to a user.

There are four ways a user can be signed in:

- [Using a built-in OAuth Provider](https://next-auth.js.org/configuration/providers/oauth) (e.g Github, Google, Okta, Apple, GitLab, Amazon, Microsoft Azure, LinkedIn, Atlassian, Auth0, etc...)
   - List of NextAuth.js providers: https://next-auth.js.org/providers/.
- [Using a custom OAuth Provider](https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider)
- [Using Email](https://next-auth.js.org/configuration/providers/email)
- [Using Credentials](https://next-auth.js.org/configuration/providers/credentials)

DiegoTavares reviewed (approved):
Pre-approved. 

Please address the comments before merging.

Zach-Fong committed:
[cueweb] Make CueWeb authentication and Sentry usage optional

- Added functionality to make login and Sentry usage optional in CueWeb.
- Updated configuration to enable/disable these features via environment variables.
- Enhanced the cueweb/README.md with instructions for configuring these options in the CueWeb setup.

Co-authored-by: Zachary Fong <zfong@imageworks.com>
Co-authored-by: Ramon Figueiredo <rfigueiredo@imageworks.com>

ramonfigueiredo committed:
[CueWeb] Update the cueweb/README.md

ramonfigueiredo commented:
@tomi-lui
